### ISSUE TITLE
systemd-nspawn: Failed to decrypt image: Invalid argument

### ISSUE THREAD

**Issue by grisia** [systemd/systemd#6525]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> systemd-234.11-3

### Used distribution

> Arch Linux
> Linux 4.9.40-1-ARCH armv6l GNU/Linux

### In case of bug report: Expected behaviour you didn't see

> Container with encrypted image boot when starting service

### In case of bug report: Unexpected behaviour you saw

> Container with encrypted image won't boot when starting service
> 
> ~ # _systemctl start systemd-nspawn@CRYPT-CONTAINER-IMAGE_

```
Please enter image passphrase! ********
Job for systemd-nspawn@CRYPT-CONTAINER-IMAGE.service failed because the control process exited with error code.
See "systemctl  status systemd-nspawn@CRYPT-CONTAINER-IMAGE.service" and "journalctl  -xe" for details.

```
>  _~ # journalctl -e_
```
~~~ snip ~~~
systemd[1]: Starting Container CRYPT-CONTAINER-IMAGE...
kernel:  loop1: p1
systemd-nspawn[19833]: Failed to decrypt image: Invalid argument
systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Main process exited, code=exited, status=1/FAILURE
systemd-udevd[19839]: inotify_add_watch(9, /dev/loop1, 10) failed: No such file or directory
systemd[1]: Failed to start Container CRYPT-CONTAINER-IMAGE.
systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Unit entered failed state.
systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Failed with result 'exit-code'.
~~~ snip ~~~
```

### In case of bug report: Steps to reproduce the problem

> Create encrypted image with ([systemd-mkosi](https://github.com/systemd/mkosi))
> _~ # mkosi_

> mkosi.default (config file for mkosi)
```
[Distribution]
Distribution=arch

[Output]
Format=raw_btrfs
Bootable=no
Encrypt=all
Output=/var/lib/machines/CRYPT-CONTAINER-IMAGE
Compress=yes
#root-size=

[Packages]
Packages=git gzip base-devel
```
> mkosi.passphrase (config file for mkosi)
```
P@ssw0rd
```

> Modifie service to work with images
>  _~ # systemctl edit  systemd-nspawn@CRYPT-CONTAINER-IMAGE_

```
[Service]
ExecStart=
ExecStart=/usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest -U --settings=override --image=/var/lib/machines/%i --machine=%i
```
> _~ # systemctl start systemd-nspawn@CRYPT-CONTAINER-IMAGE_

### Note

> Booting container direct with _systemd-nspawn_ works
> _~ # systemd-nspawn --boot --image=/var/lib/machines/CRYPT-CONTAINER-IMAGE --machine=CRYPT-CONTAINER-IMAGE_
```
Please enter image passphrase! *******
Spawning container CRYPT-CONTAINER-IMAGE on /var/lib/machines/CRYPT-CONTAINER-IMAGE.
Press ^] three times within 1s to kill container.
systemd 234 running in system mode. (+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 -SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN default-hierarchy=hybrid)
Detected virtualization systemd-nspawn.
Detected architecture arm.

Welcome to Arch Linux ARM!
~~~ snip ~~~
```

**Comment by poettering** [comment]
hmm, invoking this from the cmdline works but from a unit doesnt? that's really strange... I am puzzled I must say... And given the kernel's log message the loopback setup actually works, but then falls appart later...

Can you add Environment=SYSTEMD_LOG_LEVEL to the service's unit file, and report back what nspawn prints then?

**Comment by grisia** [comment]
_~ # systemd-analyze set-log-level debug_
_~ # systemctl edit systemd-nspawn@CRYPT-CONTAINER-IMAGE.service_

```
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
ExecStart=
ExecStart=/usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest -U --settings=override --image=/var/lib/machines/%i.raw --machine=%i
```

Just to mention here is the original service file (maybe some privileges/capabilities aren't enough?):

```
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=Container %i
Documentation=man:systemd-nspawn(1)
PartOf=machines.target
Before=machines.target
After=network.target systemd-resolved.service
RequiresMountsFor=/var/lib/machines

[Service]
ExecStart=/usr/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --network-veth -U --settings=override --machine=%i
KillMode=mixed
Type=notify
RestartForceExitStatus=133
SuccessExitStatus=133
Slice=machine.slice
Delegate=yes
TasksMax=16384

DevicePolicy=closed
DeviceAllow=/dev/net/tun rwm
DeviceAllow=char-pts rw

# nspawn itself needs access to /dev/loop-control and /dev/loop, to
# implement the --image= option. Add these here, too.
DeviceAllow=/dev/loop-control rw
DeviceAllow=block-loop rw
DeviceAllow=block-blkext rw

[Install]
WantedBy=machines.target

```

_~ # systemctl start systemd-nspawn@CRYPT-CONTAINER-IMAGE_ 

```
Broadcast message from root@alarmpi (Mon 2017-08-07 11:57:10 CEST):

Password entry required for 'Please enter image passphrase!' (PID 2493).
Please enter password with the systemd-tty-ask-password-agent tool!

Please enter image passphrase! *******
```

_~ # journalctl -ex_

```
~~~~ snip ~~~~

Aug 07 11:57:01 alarmpi systemd[1]: Accepted new private connection.
Aug 07 11:57:01 alarmpi systemd[1]: systemd-journald.service: Got notification message from PID 182 (WATCHDOG=1)
Aug 07 11:57:01 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Trying to enqueue job systemd-nspawn@CRYPT-CONTAINER-IMAGE.service/start/replace
Aug 07 11:57:01 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Installed new job systemd-nspawn@CRYPT-CONTAINER-IMAGE.service/start as 1642
Aug 07 11:57:01 alarmpi systemd[1]: var-lib-machines.mount: Installed new job var-lib-machines.mount/start as 1703
Aug 07 11:57:01 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Enqueued job systemd-nspawn@CRYPT-CONTAINER-IMAGE.service/start as 1642
Aug 07 11:57:02 alarmpi systemd[1]: var-lib-machines.mount: ConditionPathExists=/var/lib/machines.raw failed.
Aug 07 11:57:02 alarmpi systemd[1]: var-lib-machines.mount: Starting requested but condition failed. Not starting unit.
Aug 07 11:57:02 alarmpi systemd[1]: var-lib-machines.mount: Job var-lib-machines.mount/start finished, result=done
Aug 07 11:57:02 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Passing 0 fds to service
Aug 07 11:57:02 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: About to execute: /usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest -U --settings=override --image=/var/lib/machines/CRYPT-CONTAINER-IMAGE.raw --machine=CRYPT-CONTAINER-IMAGE
Aug 07 11:57:02 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Forked /usr/bin/systemd-nspawn as 2374
Aug 07 11:57:02 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Changed failed -> start
Aug 07 11:57:02 alarmpi systemd[1]: Starting Container CRYPT-CONTAINER-IMAGE...
-- Subject: Unit systemd-nspawn@CRYPT-CONTAINER-IMAGE.service has begun start-up
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit systemd-nspawn@CRYPT-CONTAINER-IMAGE.service has begun starting up.
Aug 07 11:57:02 alarmpi systemd[2374]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Executing: /usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest -U --settings=override --image=/var/lib/machines/CRYPT-CONTAINER-IMAGE.raw --machine=CRYPT-CONTAINER-IMAGE
Aug 07 11:57:02 alarmpi systemd[1]: systemd-journald.service: Got notification message from PID 182 (FDSTORE=1)
Aug 07 11:57:02 alarmpi systemd[1]: systemd-journald.service: Added fd 54 (n/a) to fd store.
Aug 07 11:57:02 alarmpi systemd[1]: systemd-udevd.service: Got notification message from PID 196 (WATCHDOG=1)
Aug 07 11:57:02 alarmpi systemd-nspawn[2374]: Settings are trusted: yes
Aug 07 11:57:02 alarmpi kernel:  loop7: p1
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.path: Got triggered.
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Trying to enqueue job systemd-ask-password-wall.service/start/replace
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Installed new job systemd-ask-password-wall.service/start as 1704
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Enqueued job systemd-ask-password-wall.service/start as 1704
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.path: Changed waiting -> running
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.path: Got triggered.
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Trying to enqueue job systemd-ask-password-wall.service/start/replace
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Merged into installed job systemd-ask-password-wall.service/start as 1704
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Enqueued job systemd-ask-password-wall.service/start as 1704
Aug 07 11:57:02 alarmpi systemd[1]: systemd-ask-password-wall.service: Job systemd-ask-password-wall.service/start finished, result=done
Aug 07 11:57:02 alarmpi systemd[1]: dev-loop7.device: Changed dead -> plugged
Aug 07 11:57:02 alarmpi systemd[1]: sys-devices-virtual-block-loop7.device: Changed dead -> plugged
Aug 07 11:57:02 alarmpi systemd[1]: dev-disk-by\x2duuid-7da2c0e8\x2d3125\x2d414b\x2d9f79\x2d655260c9ee1b.device: Changed dead -> plugged
Aug 07 11:57:02 alarmpi systemd[1]: dev-disk-by\x2dpartuuid-a527d490\x2da745\x2d486a\x2d8d61\x2d233271e798da.device: Changed dead -> plugged
Aug 07 11:57:02 alarmpi systemd[1]: dev-disk-by\x2dpartlabel-Root\x5cx20Partition.device: Changed dead -> plugged
Aug 07 11:57:02 alarmpi systemd[1]: dev-loop7p1.device: Changed dead -> plugged
Aug 07 11:57:02 alarmpi systemd[1]: sys-devices-virtual-block-loop7-loop7p1.device: Changed dead -> plugged
Aug 07 11:57:10 alarmpi systemd[1]: systemd-timesyncd.service: Got notification message from PID 215 (WATCHDOG=1)
Aug 07 11:57:10 alarmpi systemd[1]: systemd-networkd.service: Got notification message from PID 199 (WATCHDOG=1)
~~~~ ENTERING PASSWORD ~~~~
Aug 07 11:57:25 alarmpi systemd-nspawn[2374]: Failed to decrypt image: Invalid argument
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Got notification message from PID 2374 (STOPPING=1, STATUS=Terminating...)
Aug 07 11:57:25 alarmpi systemd[1]: systemd-logind.service: Got notification message from PID 223 (WATCHDOG=1)
Aug 07 11:57:25 alarmpi systemd[1]: systemd-machined.service: Got notification message from PID 392 (WATCHDOG=1)
Aug 07 11:57:25 alarmpi systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 54 (stored), closing.
Aug 07 11:57:25 alarmpi systemd[1]: Received SIGCHLD from PID 2374 (systemd-nspawn).
Aug 07 11:57:25 alarmpi systemd[1]: Child 2374 (systemd-nspawn) died (code=exited, status=1/FAILURE)
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Child 2374 belongs to systemd-nspawn@CRYPT-CONTAINER-IMAGE.service
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Main process exited, code=exited, status=1/FAILURE
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Changed start -> failed
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Job systemd-nspawn@CRYPT-CONTAINER-IMAGE.service/start finished, result=failed
Aug 07 11:57:25 alarmpi systemd[1]: Failed to start Container CRYPT-CONTAINER-IMAGE.
-- Subject: Unit systemd-nspawn@CRYPT-CONTAINER-IMAGE.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit systemd-nspawn@CRYPT-CONTAINER-IMAGE.service has failed.
--
-- The result is failed.
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Unit entered failed state.
Aug 07 11:57:25 alarmpi systemd[1]: systemd-nspawn@CRYPT-CONTAINER-IMAGE.service: Failed with result 'exit-code'.
Aug 07 11:57:25 alarmpi systemd[1]: Got disconnect on private connection.
Aug 07 11:57:25 alarmpi systemd-udevd[2380]: inotify_add_watch(9, /dev/loop7, 10) failed: No such file or directory
Aug 07 11:57:25 alarmpi systemd[1]: dev-loop7.device: Changed plugged -> dead
Aug 07 11:57:25 alarmpi systemd[1]: dev-loop7.device: Collecting.
Aug 07 11:57:25 alarmpi systemd[1]: dev-disk-by\x2duuid-7da2c0e8\x2d3125\x2d414b\x2d9f79\x2d655260c9ee1b.device: Changed plugged -> dead
Aug 07 11:57:25 alarmpi systemd[1]: dev-disk-by\x2duuid-7da2c0e8\x2d3125\x2d414b\x2d9f79\x2d655260c9ee1b.device: Collecting.
Aug 07 11:57:25 alarmpi systemd[1]: sys-devices-virtual-block-loop7.device: Changed plugged -> dead
Aug 07 11:57:25 alarmpi systemd[1]: sys-devices-virtual-block-loop7.device: Collecting.

```

**Comment by poettering** [comment]
Next step would be to strace, to see which syscall precisely generates EINVAL... Can you prefix the ExecStart= line with "/usr/bin/strace -f -s500 -o /tmp/nspawn.log …" so that we can follow what nspawn is doing and where precisely it fails?

**Comment by poettering** [comment]
(my educated guess is that we need some more DeviceAllow= lines here for the DM devices used on your system)

**Comment by grisia** [comment]
Great that's it. Thx for the hint...

For me to get encrypted images working I had to add the following lines to my service file:

```
DeviceAllow=/dev/mapper/control
DeviceAllow=block-device-mapper
```

Adding _... /usr/bin/strace -f -s500 -o /tmp/nspawn.log ..._ to my ExecStart-Line shows me permission errors on _/dev/mapper/control_ and _/dev/mapper/loop1p1_.

Because of I don't which loop[1..n] device will be taken, I took block-device-mapper.

May be printing these permission errors to systemlog would help to get in the right direction...

**Comment by poettering** [comment]
Fix waiting in #6580
